js/private/js_run_devserver.mjs

```diff
@@ -447,38 +447,44 @@ async function main(args, sandbox) {
             process.exit(code)
         })
 
+        // Process stdin data in order using a promise chain.
         let syncing = Promise.resolve()
         process.stdin.on('data', async (chunk) => {
+            return (syncing = syncing.then(() => processChunk(chunk)))
+        })
+
+        async function processChunk(chunk) {
             try {
                 const chunkString = chunk.toString()
                 if (chunkString.includes('IBAZEL_BUILD_COMPLETED SUCCESS')) {
                     if (process.env.JS_BINARY__LOG_DEBUG) {
                         console.error('IBAZEL_BUILD_COMPLETED SUCCESS')
                     }
-                    // Chain promises via syncing.then()
-                    syncing = syncing.then(() => {
-                        // Re-parse the config file to get the latest list of data files to copy
-                        const updatedDataFiles = JSON.parse(
-                            fs.readFileSync(configPath)
-                        ).data_files
+
+                    const oldFiles = config.data_files
+
+                    // Re-parse the config file to get the latest list of data files to copy
+                    const updatedDataFiles = JSON.parse(
+                        fs.readFileSync(configPath)
+                    ).data_files
+
+                    // Await promises to catch any exceptions, and wait for the
+                    // sync to be complete before writing to stdin of the child
+                    // process
+                    await Promise.all([
                         // Remove files that were previously synced but are no longer in the updated list of files to sync
-                        deleteFiles(
-                            config.data_files,
-                            updatedDataFiles,
-                            sandbox
-                        )
+                        deleteFiles(oldFiles, updatedDataFiles, sandbox),
```

Comment:
Note that this returns a promise, which this PR has not changed, but previously we were never doing `await` anywhere. So the `deleteFiles` and `syncFiles` were running async and concurrently (and they still are), but nothing was doing `await` to wait for them.

This means the `await syncing` below was essentially not waiting for anything. Then this code could start to be executed multiple times in parallel, defeating hte purpose of `syncing`...